discovery-provider/src/queries/get_premium_track_signatures.py

```diff
@@ -0,0 +1,320 @@
+                return True
+        except Exception as e:
+            logger.error(
+                f"Could not get nft balance for erc721 nft collection {contract_address} and user wallet {wallet}. Error: {e}"
+            )
+    return False
+
+
+def _does_user_own_erc1155_nft_collection(
+    contract_address: str, user_eth_wallets: List[ChecksumAddress], token_ids: List[int]
+):
+    for wallet in user_eth_wallets:
+        try:
+            contract = eth_web3.eth.contract(address=contract_address, abi=erc1155_abi)
+            nft_balances = contract.functions.balanceOfBatch(
+                [wallet] * len(token_ids), token_ids
+            ).call()
+            positive_nft_balances = list(
+                filter(lambda nft_balance: int(nft_balance) > 0, nft_balances)
+            )
+            if len(positive_nft_balances) > 0:
+                return True
+        except Exception as e:
+            logger.error(
+                f"Could not get nft balance for erc1155 nft collection {contract_address} and user wallet {wallet}. Error: {e}"
+            )
+    return False
+
+
+# Returns premium tracks from given track ids with an nft collection as the premium conditions.
+def _get_nft_gated_tracks(track_ids: List[int], session: Session):
+    premium_tracks = (
+        session.query(Track)
+        .filter(
+            Track.is_premium == True,
+            Track.track_id.in_(track_ids),
+            Track.is_current == True,
+            Track.is_delete == False,
+        )
+        .all()
+    )
+    nft_gated_tracks = list(
+        filter(lambda track: "nft_collection" in track, premium_tracks)
+    )
+    return nft_gated_tracks
+
+
+def _get_eth_nft_gated_track_signatures(
+    user_wallet: str,
+    eth_associated_wallets: List[str],
+    tracks: List[Track],
+    track_token_id_map: Dict[int, List[str]],
+):
+    track_signature_map = {}
+
+    user_eth_wallets = list(
+        map(Web3.toChecksumAddress, eth_associated_wallets + [user_wallet])
+    )
+
+    # Handle premium tracks gated by ERC721 nft collections
+    erc721_gated_tracks = list(
+        filter(
+            lambda track: track.premium_conditions["nft_collection"]["standard"]
+            == "ERC721",
+            tracks,
+        )
+    )
+
+    # Build a map of ERC721 collection address -> track ids
+    # so that only one chain call will be made for premium tracks
+    # that share the same nft collection gate.
+    erc721_collection_track_map = defaultdict(list)
+    for track in erc721_gated_tracks:
+        contract_address = Web3.toChecksumAddress(
+            track.premium_conditions["nft_collection"]["address"]
+        )
+        erc721_collection_track_map[contract_address].append(track.track_id)
+
+    # Check ownership of nfts from erc721 collections from given contract addresses,
+    # using all user eth wallets, and generate signatures for corresponding tracks.
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
Curious what folks like @isaacsolo think, but IMO we should continue moving away from the threadpool executor model of concurrency and use asyncio instead, which is much simpler & made for IO bound tasks. That said, I don't think version 1 here even needs any of this concurrency complexity at all. I'd say we see how much we end up requesting batch tracks, and then optimize as needed. Do we have any ballpark perf numbers for how long it takes to fetch 1 / multiple NFTs from this endpoint?

Reply:
yeah down to chat what direction we want to go. i think if we can already build the concurrency then might as well.
that being said, i tested with 2 nfts, 1 erc721 and 1 erc115, _before_ adding concurrency, took about a second. but can get you _exact_ time elapsed today :)

Reply:
yup synced with salliou earlier! this looks like a good candidate for asyncio but if it's a cheap query nbd

Reply:
since nbd either way rn, my take is to leave the concurrency in the code as-is since it's been tested and it works. also haven't really dug into how well asyncio will work with web3.
if ok with y'all, we just proceed with what's implemented here :) and get to other juicy premium content things

Reply:
sgtm 👍